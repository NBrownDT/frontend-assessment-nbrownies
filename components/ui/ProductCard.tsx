import React from "react";

import { formatPrice } from "@utils/formatPrice";

import styles from "./ProductCard.module.css";

type ProductCardProps = {
  images?: string[];
  priceInPence: number;
  reviews?: any; // TODO - don't use any
  title: string;
};

const ProductCard: React.FC<ProductCardProps> = ({
  images,
  priceInPence,
  reviews,
  title,
}) => {
  const priceString = formatPrice(priceInPence);
  // TODO: remove null check, some of the image arrays are undefined
  const heroImage = `/images/products/${images?.[0]}`;
  return (
    <article className={styles.Root}>
      <figure className={styles.figure}>
        <img
          src={heroImage}
          alt={title}
          width={240}
          height={240}
          className={styles.heroImage}
        />
      </figure>
      <h3 className={styles.title}>{title}</h3>
      <div className={styles.meta}>
        <p className={styles.price}>{priceString}</p>
        {!!reviews && (
          <div className={styles.reviewsContainer}>
            <p className={styles.reviewsDetail}>
              <span className={styles.reviewsAverage}>★★★★☆</span>
              {/* <span className={styles.reviewsAverage}>{reviews.average}</span> */}
              <span className={styles.reviewsCount}>({reviews.count})</span>
            </p>
          </div>
        )}
      </div>
    </article>
  );
};

export default ProductCard;
