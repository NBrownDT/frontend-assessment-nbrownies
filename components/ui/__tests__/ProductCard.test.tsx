import React from "react";
import "@testing-library/jest-dom";
import { render } from "@testing-library/react";
import ProductCard from "../ProductCard";

const sampleProductsDietary = {
  vegan: false,
  wheatFree: false,
  halal: false,
  nutFree: false,
  glutenFree: false,
};

const sampleProductsReviewSummary = {
  count: 42,
  average: 4.2,
};

const mockProduct = {
  // title: "Birthday Giftbox with Coffee",
  title: "Onboarding Giftbox with Strong Coffee",
  images: ["brownies-birthday-coffee.webp"],
  priceInPence: true,
  newIn: false,
  nutrition: {
    kcal: 100,
    carbs: 10,
  },
  dietary: { ...sampleProductsDietary },
  reviewsSummary: { ...sampleProductsReviewSummary },
};

describe("ProductCard tests", () => {
  it("renders the ProductCard correctly", () => {
    const { container } = render(
      <ProductCard
        images={mockProduct.images}
        priceInPence={mockProduct.priceInPence}
        reviews={mockProduct.reviewsSummary}
        title={mockProduct.title}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
