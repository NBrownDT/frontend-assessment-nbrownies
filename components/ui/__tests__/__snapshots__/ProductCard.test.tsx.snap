// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard tests renders the ProductCard correctly 1`] = `
<div>
  <article
    class="Root"
  >
    <figure
      class="figure"
    >
      <img
        alt="Onboarding Giftbox with Strong Coffee"
        class="heroImage"
        data-nimg="1"
        decoding="async"
        height="240"
        loading="lazy"
        src="/_next/image?url=%2Fimages%2Fproducts%2Fbrownies-birthday-coffee.webp&w=640&q=75"
        srcset="/_next/image?url=%2Fimages%2Fproducts%2Fbrownies-birthday-coffee.webp&w=256&q=75 1x, /_next/image?url=%2Fimages%2Fproducts%2Fbrownies-birthday-coffee.webp&w=640&q=75 2x"
        style="color: transparent;"
        width="240"
      />
    </figure>
    <h3
      class="title"
    >
      Onboarding Giftbox with Strong Coffee
    </h3>
    <div
      class="meta"
    >
      <p
        class="price"
      >
        £0.01
      </p>
      <div
        class="reviewsContainer"
      >
        <p
          class="reviewsDetail"
        >
          <span
            class="reviewsAverage"
          >
            ★★★★☆
          </span>
          <span
            class="reviewsCount"
          >
            (
            42
            )
          </span>
        </p>
      </div>
    </div>
  </article>
</div>
`;
